extensions/positron-reticulate/src/extension.ts

```diff
@@ -0,0 +1,484 @@
+		// tied to the R session.
+		// We have to send a restart to the R session, and send a reticulate::repl_python()
+		// command to it.
+		this.pythonSession.shutdown(positron.RuntimeExitReason.Restart);
+		await this.rSession.restart();
+		const rSession = await getRSession();
+		rSession.execute(
+			'reticulate::repl_python()',
+			'start-reticulate',
+			positron.RuntimeCodeExecutionMode.Interactive,
+			positron.RuntimeErrorBehavior.Continue
+		);
+		return;
+	}
+
+	public async shutdown() {
+		await this.pythonSession.shutdown(positron.RuntimeExitReason.Shutdown);
+		// Tell Positron that the kernel has exit. When launching IPykernel from a standalone
+		// process, when the kernel exits, then all of it's threads, specially the IOPub thread
+		// holding the ZeroMQ sockets will cease to exist, and thus Positron identifies that the
+		// kernel has successfuly closed. However, since we launch positron from a different thread,
+		// when the kernel exits, the thread exits, but all other dangling threads are still alive,
+		// thus Positron never identifies that the kernel exited. We must then manually fire exit event.
+		// We rely on an implementation detail of the jupyter adapter, that allows us to force the
+		// kernels to disconnect.
+		(this.pythonSession as any)._kernel._kernel._allSockets.forEach((socket: any) => socket.disconnect());
+		return;
+	}
+
+	public forceQuit() {
+		// Force quit will kill the process, which also kills the R process.
+		return this.pythonSession.forceQuit();
+	}
+
+	public dispose() {
+		return this.pythonSession.dispose();
+	}
+}
+
+async function getRSession(): Promise<positron.LanguageRuntimeSession> {
+
+	// Retry logic to start an R session.
+	const maxRetries = 5;
+	let session;
+	let error;
+	for (let i = 0; i < maxRetries; i++) {
+		try {
+			session = await getRSession_();
+		}
+		catch (err) {
+			console.warn(`Could not find an R session .Retrying(${i} / ${maxRetries}): ${err} `);
+			error = err; // Keep the last error so we can display it
+		}
+	}
+
+	if (!session) {
+		throw new Error(`Could not initialize an R session to launch reticulate. ${error}`);
+	}
+
+	// Check that we have a minimum version of reticulate.
+	if (!await session.callMethod?.('is_installed', 'reticulate', '1.39')) {
+		throw new Error('Reticulate >= 1.39 is required');
+	}
+
+	return session;
+}
+
+async function getRSession_(): Promise<positron.LanguageRuntimeSession> {
+	let session = await positron.runtime.getForegroundSession();
+	if (!session || session.runtimeMetadata.languageId !== 'r') {
+		const runtime = await positron.runtime.getPreferredRuntime('r');
+		await positron.runtime.selectLanguageRuntime(runtime.runtimeId);
+		session = await positron.runtime.getForegroundSession();
+	}
+
+	if (!session) {
+		throw new Error(`No available R session to execute reticulate`);
+	}
+
+	return session;
+}
+
+async function* reticulateRuntimesDiscoverer() {
+	const runtimeMetadata = new ReticulateRuntimeMetadata();
```

Comment:
As written *everyone* is going to see a "reticulate" interpreter available, even if they don't have R or Python installed (ironically users who have a bare machine with no R or Python will see, as the only option, something that requires both!). I don't think we want that; the other interpreters only show up if they're actually installed.

I think we should ensure that only people with some understanding of reticulate see this option since the behavior is pretty confusing if you don't understand the internals (no fault of this change, we just don't have the UX to communicate clearly the nested relationship of the R and Python sessions). We could do that a few ways:

- make this opt-in; you need to go to the Settings to turn on reticulate (after which everything will work as it does now), or
- detect users who are already reticulate users and only show the option to them (e.g. don't register the runtime until we have a suitable R session with the reticulate package installed, after which we can light it up), or
- some other way?

Reply:
I agree; we should only show the reticulated-Python interpreter if some basic criteria are met. Perhaps the criterion could be "Is R installed?"

If R is installed, then we could offer a few convenience dialogues, similar to the way we offer to install ipykernel:

- Offer to install reticulate (via `install.packages("reticulate")`)
- Offer to install Python (using `reticulate::install_python()`, which uses pyenv to install Python)
- Offer to configure a default virtual environment for reticulate (using `reticulate::virtualenv_create("r-reticulate")`)